mender-update/standalone/standalone.cpp

```diff
@@ -0,0 +1,498 @@
+	}
+
+	exp_array = json::Get<vector<string>>(json, keys.payload_types, json::MissingOk::No);
+	if (!exp_array) {
+		return expected::unexpected(exp_array.error());
+	}
+	dst.payload_types = exp_array.value();
+
+	if (dst.version != context::MenderContext::standalone_data_version) {
+		return expected::unexpected(error::Error(
+			make_error_condition(errc::not_supported),
+			"State data has a version which is not supported by this client"));
+	}
+
+	if (dst.artifact_name == "") {
+		return expected::unexpected(context::MakeError(
+			context::DatabaseValueError, "`" + keys.artifact_name + "` is empty"));
+	}
+
+	if (dst.payload_types.size() == 0) {
+		return expected::unexpected(context::MakeError(
+			context::DatabaseValueError, "`" + keys.payload_types + "` is empty"));
+	}
+	if (dst.payload_types.size() >= 2) {
+		return expected::unexpected(error::Error(
+			make_error_condition(errc::not_supported),
+			"`" + keys.payload_types + "` contains multiple payloads"));
+	}
+
+	return dst;
+}
+
+void DataFromPayloadHeaderView(const artifact::PayloadHeaderView &header, Data &dst) {
+	dst.version = context::MenderContext::standalone_data_version;
+	dst.artifact_name = header.header.artifact_name;
+	dst.artifact_group = header.header.artifact_group;
+	dst.artifact_provides = header.header.type_info.artifact_provides;
+	dst.artifact_clears_provides = header.header.type_info.clears_artifact_provides;
+	dst.payload_types.clear();
+	dst.payload_types.push_back(header.header.payload_type);
+}
+
+error::Error SaveData(database::KeyValueDatabase &db, const Data &data) {
+	DataKeys keys;
+	stringstream ss;
+	ss << "{";
+	ss << "\"" << keys.version << "\":" << data.version;
+
+	ss << ",";
+	ss << "\"" << keys.artifact_name << "\":\"" << data.artifact_name << "\"";
+
+	ss << ",";
+	ss << "\"" << keys.artifact_group << "\":\"" << data.artifact_group << "\"";
+
+	ss << ",";
+	ss << "\"" << keys.payload_types << "\": [";
+	bool first = true;
+	for (auto elem : data.payload_types) {
+		if (!first) {
+			ss << ",";
+		}
+		ss << "\"" << elem << "\"";
+		first = false;
+	}
+	ss << "]";
+
+	if (data.artifact_provides) {
+		ss << ",";
+		ss << "\"" << keys.artifact_provides << "\": {";
+		bool first = true;
+		for (auto elem : data.artifact_provides.value()) {
+			if (!first) {
+				ss << ",";
+			}
+			ss << "\"" << elem.first << "\":\"" << elem.second << "\"";
+			first = false;
+		}
+		ss << "}";
+	}
+
+	if (data.artifact_clears_provides) {
+		ss << ",";
+		ss << "\"" << keys.artifact_clears_provides << "\": [";
+		bool first = true;
+		for (auto elem : data.artifact_clears_provides.value()) {
+			if (!first) {
+				ss << ",";
+			}
+			ss << "\"" << elem << "\"";
+			first = false;
+		}
+		ss << "]";
+	}
+
+	ss << "}";
+
+	string strdata = ss.str();
+	vector<uint8_t> bytedata(strdata.begin(), strdata.end());
+
+	return db.Write(context::MenderContext::standalone_state_key, bytedata);
+}
+
+error::Error RemoveData(database::KeyValueDatabase &db) {
```

Comment:
*RemoveStateData()*?

Reply:
StateData, of course! @oleorhagen were discussing what to call it, and we didn't think of this simple name! 😂